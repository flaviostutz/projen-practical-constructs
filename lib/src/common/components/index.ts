export * from './base-tooling';
export * from './common-target-type';
export * from './common-targets';
export * from './release-tasks';
export * from './publish';
