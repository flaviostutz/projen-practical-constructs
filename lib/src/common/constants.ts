export const PROJEN_VERSION = '0.91.13';
export const NODE_VERSION = '20.16.0';
export const MONOTAG_VERSION = '1.23.0';
