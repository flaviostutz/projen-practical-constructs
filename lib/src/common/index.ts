export * from './components';
export * from './files';
export * from './types';
