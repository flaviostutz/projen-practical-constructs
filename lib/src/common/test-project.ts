import { mkdtempSync } from 'node:fs';
import { tmpdir } from 'node:os';
// eslint-disable-next-line unicorn/import-style
import { join } from 'node:path';

import { Project } from 'projen';

export class TestProject extends Project {
  constructor() {
    const outdir = mkdtempSync(join(tmpdir(), 'test-'));
    super({ name: 'test', outdir });
  }
}
