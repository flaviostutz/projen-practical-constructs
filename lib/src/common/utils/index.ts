export * from './ts2jsii';
export * from './monotag';
