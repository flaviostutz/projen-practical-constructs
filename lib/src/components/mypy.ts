/* eslint-disable no-new */
import { Component, DependencyType, Project } from 'projen';

import { MyPyIniFile } from '../files';
import { addTaskToParent, TaskOptions } from '../tasks';

export class MyPy extends Component {
  constructor(project: Project, taskOpts: TaskOptions) {
    super(project);

    new MyPyIniFile(project);
    project.deps.addDependency('mypy@1.14.*', DependencyType.DEVENV);
    project.addGitIgnore('.mypy_cache');
    const lintMypyTask = project.tasks.addTask('lint-mypy', {
      description: `Code type checks (mypy)`,
      exec: `${taskOpts.venvPath}/bin/mypy src tests`,
    });
    addTaskToParent(project, lintMypyTask, taskOpts.attachTasksTo);
  }
}
