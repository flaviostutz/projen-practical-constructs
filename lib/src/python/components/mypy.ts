/* eslint-disable no-new */
import { Component, DependencyType, Project } from 'projen';

import { MyPyIniFile } from '../files';
import { addSpawnTaskToExisting, TaskOptions } from '../tasks';
import { CommonTargets } from '../../common/components/common-target-type';

export class MyPy extends Component {
  constructor(project: Project, taskOpts: TaskOptions) {
    super(project);

    new MyPyIniFile(project);
    project.deps.addDependency('mypy@1.14.*', DependencyType.DEVENV);
    project.addGitIgnore('.mypy_cache');
    const lintMypyTask = project.tasks.addTask('lint-mypy', {
      description: `Code type checks (mypy)`,
      exec: `${taskOpts.venvPath}/bin/mypy src tests`,
    });
    addSpawnTaskToExisting(project, lintMypyTask, CommonTargets.LINT);
  }
}
