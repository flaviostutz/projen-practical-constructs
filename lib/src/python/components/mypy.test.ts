/* eslint-disable no-new */
import { Testing } from 'projen';

import { TestProject } from '../../common/test-project';

import { MyPy } from './mypy';

describe('MyPy', () => {
  describe('MyPy', () => {
    test('snapshot of default options', () => {
      const project = new TestProject();
      new MyPy(project, { venvPath: '.venv' });
      const out = Testing.synth(project);
      expect(out).toMatchSnapshot();
    });

    test('custom venvPath', () => {
      const project = new TestProject();
      new MyPy(project, { venvPath: './custom_venv' });
      const out = Testing.synth(project);
      expect(out).toMatchSnapshot();
    });

    test('attachTasksTo option', () => {
      const project = new TestProject();
      project.addTask('build1', {
        exec: 'echo "Building project..."',
      });
      new MyPy(project, { venvPath: '.venv', attachTasksTo: 'build1' });
      const out = Testing.synth(project);
      expect(out).toMatchSnapshot();
    });

    test('without attachTasksTo option', () => {
      const project = new TestProject();
      new MyPy(project, { venvPath: '.venv' });
      const out = Testing.synth(project);
      expect(out).toMatchSnapshot();
    });
  });
});
