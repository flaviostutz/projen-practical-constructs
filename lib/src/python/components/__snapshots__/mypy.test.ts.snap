// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyPy MyPy attachTasksTo option 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/mypy.ini
.mypy_cache
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "dependencies": [
      {
        "name": "mypy",
        "type": "devenv",
        "version": "1.14.*",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "mypy.ini",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "build1": {
        "name": "build1",
        "steps": [
          {
            "exec": "echo "Building project..."",
          },
          {
            "spawn": "lint-mypy",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "lint-mypy": {
        "description": "Code type checks (mypy)",
        "name": "lint-mypy",
        "steps": [
          {
            "exec": ".venv/bin/mypy src tests",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "mypy.ini": "[mypy]
mypy_path = 
    src
    tests
warn_return_any = True
warn_unreachable = True
warn_redundant_casts = True
warn_unused_ignores = True
pretty = True
cache_dir = .cache/.mypy_cache
",
}
`;

exports[`MyPy MyPy custom venvPath 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/mypy.ini
.mypy_cache
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "dependencies": [
      {
        "name": "mypy",
        "type": "devenv",
        "version": "1.14.*",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "mypy.ini",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "lint-mypy": {
        "description": "Code type checks (mypy)",
        "name": "lint-mypy",
        "steps": [
          {
            "exec": "./custom_venv/bin/mypy src tests",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "mypy.ini": "[mypy]
mypy_path = 
    src
    tests
warn_return_any = True
warn_unreachable = True
warn_redundant_casts = True
warn_unused_ignores = True
pretty = True
cache_dir = .cache/.mypy_cache
",
}
`;

exports[`MyPy MyPy snapshot of default options 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/mypy.ini
.mypy_cache
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "dependencies": [
      {
        "name": "mypy",
        "type": "devenv",
        "version": "1.14.*",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "mypy.ini",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "lint-mypy": {
        "description": "Code type checks (mypy)",
        "name": "lint-mypy",
        "steps": [
          {
            "exec": ".venv/bin/mypy src tests",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "mypy.ini": "[mypy]
mypy_path = 
    src
    tests
warn_return_any = True
warn_unreachable = True
warn_redundant_casts = True
warn_unused_ignores = True
pretty = True
cache_dir = .cache/.mypy_cache
",
}
`;

exports[`MyPy MyPy without attachTasksTo option 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/mypy.ini
.mypy_cache
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "dependencies": [
      {
        "name": "mypy",
        "type": "devenv",
        "version": "1.14.*",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "mypy.ini",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "lint-mypy": {
        "description": "Code type checks (mypy)",
        "name": "lint-mypy",
        "steps": [
          {
            "exec": ".venv/bin/mypy src tests",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "mypy.ini": "[mypy]
mypy_path = 
    src
    tests
warn_return_any = True
warn_unreachable = True
warn_redundant_casts = True
warn_unused_ignores = True
pretty = True
cache_dir = .cache/.mypy_cache
",
}
`;
